### PR TITLE
[4055] Change QA and Staging to use the service domain

### PR DESCRIPTION
**Pull request by gpeng**
### Context

We are moving Register to a service domain. Initially we wanted to serve the app on both the old education and new service domains, get everything working and then add a redirect from the old to the new.

We have implemented the Signin integration with an explicitly configured URL so we can only have one or the other with the current implementation. I have spiked changing it to be dynamic but as far as I can tell it would involve a custom strategy to implement the redirect/callback and logout URLs cleanly and that's a significant change to our authentication (which has already been pen tested).

We also use the base URL elsewhere (e.g. trainee URL in CSV downloads) which would also need to be updated to handle the URL dynamically.

This PR just makes the change in QA and Staging and adds a redirect which is a copy of the Apply implementation (excluding production for now so as we can test on staging).

### Changes proposed in this pull request

* Add a redirect in rack to redirect .education.gov.uk to .service.gov.uk
* Update the base_url for staging and qa to the service one.

### Guidance to review

The implementation of the redirect is already in use in production with Apply. 

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

**Comment by misaka** [comment]
I think the deploy is failing because of this check:

```ruby
    unless application_config["application_uris"].include?(Settings.base_url)
      raise "base url #{Settings.base_url} is not in VCAP_APPLICATION['application_uris']"
    end
```
https://github.com/DFE-Digital/register-trainee-teachers/pull/2293/files#diff-1eb4f6ad7938bc433b6705c323a2461099d36986394acae217a772cc001ad678R6-R8

My guess is that `Settings.base_url`, which would be `https://qa.register-trainee-teachers.service.gov.uk` in QA, and  is not in `VCAP_APPLICATION['application_uris']`. Shame we don't get that message in the logs to make it clear.

The Right:tm: way to fix this would be to set `SETTINGS__BASE_URL=https://register-pr-2293.london.cloudapps.digital` for the deploy.

**Comment by gpeng** [comment]
[Check service domain on staging](https://trello.com/c/sHA9C7Yq/4055-check-service-domain-on-staging)

**Comment by misaka** [comment]
Nothing like reviewing your own code. :100:

File: config/initializers/canonical_rails.rb [hunk 1]
```diff
@@ -9,8 +9,8 @@
 
   # This is the main host, not just the TLD, omit slashes and protocol. If you have more than one, pick the one you want to rank in search results.
 
-  config.host = ENV.fetch("APPLICATION_HOST", "www.register-trainee-teachers.education.gov.uk")
-  config.port # = '3000'
+  config.host = ENV.fetch("APPLICATION_HOST", URI(Settings.base_url).host)
```
[the remark this answers is not in the document]
Comment (misaka) [reply config/initializers/canonical_rails.rb:12]: This isn't host verification, this is just for `CanonicalRails` which is used when generating the HTML:  

https://github.com/DFE-Digital/register-trainee-teachers/blob/f32478e701f00dcb2b8a116e39c40fe7e8618bb1/app/views/layouts/application.html.erb#L7

We aren't using hosts authorisation any more, that was removed after we saw how it complicated things. I'll remove the misleading comments here [`74ac9a8` (#2293)](https://github.com/DFE-Digital/register-trainee-teachers/pull/2293/commits/74ac9a85a5624d61e6bb0909b6336b97cd0e081a#diff-1eb4f6ad7938bc433b6705c323a2461099d36986394acae217a772cc001ad678R21) and further down in that file. Are there any others?

File: terraform/modules/paas/variables.tf [hunk 2]
```diff
@@ -44,7 +44,7 @@ locals {
   app_environment = merge(var.app_config_variable, var.app_secrets_variable, {
     DATABASE_URL  = cloudfoundry_service_key.postgres-key.credentials.uri
     SETTINGS__BLAZER_DATABASE_URL = cloudfoundry_service_key.postgres-blazer-key.credentials.uri
-  })
+  }, var.app_environment == "review" ? { SETTINGS__BASE_URL = "https://${local.web_app_name}.london.cloudapps.digital" } : {})
```
[the remark this answers is not in the document]
Comment (misaka) [reply terraform/modules/paas/variables.tf:47]: 👍 

Does that mean into the `locals` section of that same file, or should it be in here:  https://github.com/DFE-Digital/register-trainee-teachers/blob/4055-check-service-domain-on-staging/terraform/modules/paas/variables.tf ?

[the remark this answers is not in the document]
Comment (misaka) [reply terraform/modules/paas/variables.tf]: Fixed

**Review by saliceti — APPROVED** [review]

**Pull request merged**